fix(docs-infra): typography, layout, and UI polish

gkalpak opened the pull request:
This PR is based on (and supercedes) #38750 (which was in turn based on #36045). It:
- Splits independent changes into separate commits.
- Addresses relevant review feedback from #38750 and #36045.
- Includes a couple of additional fixes/improvements.

The PR contains the following changes (see individual commits for details):
- **[fix]** Update the sidenav background to better distinguish from main content.
- **[fix]** Remove uppercase style from sidenav nav-items.
- **[fix]** Fix padding in sidenav and ensure all nav-items have equal height.
- **[fix]** Improve heading styles.
- **[fix]** Make content links more prominent on hover.
- **[fix]** Improve typography styles (incl. bigger font size/line height).
- **[fix]** Improve legibility and focus on content.
- **[refactor]** Clean up styles in `_card.scss`.
- **[fix]** Fix the styling of the cards in docs introduction page.
- **[refactor]** Simplify the CSS rules for styling the "collapsed/expanded" sidenav icons.
- **[refactor]** Remove unused styles for `.heading` CSS class.

mary-poppins commented:
You can preview c4df61e at https://pr40427-c4df61e.ngbuilds.io/.

mary-poppins commented:
You can preview 65ec9c9 at https://pr40427-65ec9c9.ngbuilds.io/.

petebacondarwin commented on aio/src/styles/2-modules/_heading-anchors.scss:
```diff
@@ -7,8 +7,9 @@
       visibility: hidden;
-      display: inline-block;
-      vertical-align: text-top;
+      display: -ms-inline-flexbox;
```
According to [this](https://caniuse.com/?search=inline-flex) we don't need `-ms-inline-flexbox` for IE11.

gkalpak replied:
Oh, I forgot we are not supporting IE<11 any more :tada:

petebacondarwin commented on aio/src/styles/0-base/_typography.scss:
```diff
@@ -107,15 +107,12 @@ a {
   }
 }
 
-p {
-  margin: 14px 0 0;
-}
-
```
I see in Chrome that there is a built in user agent style of:

```css
margin-block-start: 1em;
margin-block-end: 1em;
```

which I guess is why this style here has been dropped.

Is that user agent styling standard across all supported browsers?

gkalpak replied:
I am not sure tbh. I ported this over from #38750, so I don't know the reasoning. (Maybe @sjtrimble remembers.)
That said, I would be very surprised if browsers do not apply some margin around `<p>` elements by default.

Or maybe we should just add `p { margin: 1em 0; }` just to be on the safe side.

gkalpak replied:
I added the `p { margin: 1em 0; }` style.

petebacondarwin commented on aio/src/styles/1-layouts/_content-layout.scss:
```diff
@@ -6,6 +6,8 @@
 .sidenav-content {
   min-height: 100vh;
   padding: 80px 3rem 2rem;
+  max-width: 50em;
+  margin: 0 auto;
```
I note that the resources and events pages don't seem to be limited to this max-width. Is that by design?

gkalpak replied:
Yes, there are some page that are intentionally expcluded from the `max-width`. (See also https://github.com/angular/angular/pull/38750#pullrequestreview-498996056.)

gkalpak replied:
BTW, the code for excluding them is [here][1].

[1]: https://github.com/angular/angular/blob/65ec9c9e5aa0768f3df87904f5fead6085d661fa/aio/src/styles/1-layouts/_content-layout.scss#L21-L30

petebacondarwin reviewed (changes requested):
Nice work @gkalpak - I loved the clean commits.
I think we should resolve the contrast issue, otherwise LGTM.

kapunahelewong commented on aio/src/styles/0-base/_typography.scss:
```diff
@@ -28,37 +29,52 @@ h1 {
+    content: '';
+    display: block;
+    border-bottom: 1px solid $lightgray;
+    margin: 4rem 0;
```
The 4rem looks good on my display, but out of curiosity, why not have the border styles on the `h2` (instead of on the `before`) like this?

```
h2 {
    ...
    border-top: 1px solid $lightgray;
    padding-top: 4rem;
    margin-top: 4rem;
}
```

gkalpak replied:
I ported this change from  #38750 and I've tried to keep the code as similar as possible (hence the `4rem` and the `:before` block). I was curious about the `:before` block as well and have it on my TODO list to investigate using border-styles instead in a subsequent PR.

gkalpak replied:
OK, this looked like a small change so I did it anyway :grin:

kapunahelewong commented on aio/src/styles/1-layouts/_sidenav.scss:
```diff
@@ -29,8 +29,8 @@ mat-sidenav.mat-sidenav.sidenav {
   min-width: 260px;
-  background-color: $white;
-  box-shadow: 6px 0 6px rgba(0,0,0,0.10);
+  background-color: $lightblue;
```
Second this. My first thought was to lighten the background, but @petebacondarwin's suggestion for a darker font color is a better idea.

gkalpak replied:
Done (changed it #444 - same as `.level-1` items).

Splaktar replied:
This background change also causes contrast issues with the selected span.

This `$blue` text here:
![Screen Shot 2021-01-20 at 00 08 18](https://user-images.githubusercontent.com/3506071/105130034-48fbb280-5ab4-11eb-9b24-2bd6775bb066.png)

![Screen Shot 2021-01-20 at 00 08 04](https://user-images.githubusercontent.com/3506071/105130037-49944900-5ab4-11eb-9a7c-6453ac3cf086.png)

Can we use a slightly darker blue for this?

gkalpak replied:
I couldn't find a blue-ish color that has enough contrast with the background and also be distinguishable enough from the gray color of unselected entries.

Therefore, I slightly changed the sidenav background to a light gray (and made a selected items' color a little darker):

![sidenav-#F2F2F2-#1669bb](https://user-images.githubusercontent.com/8604205/106157059-f8a6e380-618a-11eb-911b-cace70db4481.png)

WDYT?

aikidave replied:
I think it looks great!

kapunahelewong commented on aio/src/styles/2-modules/_card.scss:
```diff
@@ -2,13 +2,14 @@
+      margin: 24px 16px;
       padding-bottom: 48px;
       position: relative;
 
```
Love this update!

kapunahelewong reviewed (approved):
This is super! Just had a couple of comments. The separated commits made this easy to review. Thank you!

gkalpak commented:
@petebacondarwin, @kapunahelewong: Thx for the reviews. I've made the changes. PTAL.

petebacondarwin commented:
Oops:

```
Error: ./src/styles/main.scss
Module build failed (from ./node_modules/mini-css-extract-plugin/dist/loader.js):
ModuleBuildError: Module build failed (from ./node_modules/sass-loader/dist/cjs.js):
SassError: unmatched "}".
    ╷
111 │ }
    │ ^
    ╵
  src/styles/0-base/_typography.scss 111:1  @import
  src/styles/0-base/_base-dir.scss 5:9      @import
  src/styles/main.scss 11:9                 root stylesheet
```

gkalpak commented:
Ooops, I was targeting the wrong commit from a fixup, which messed up my rebase a bit 😇
Fixed now!

mary-poppins commented:
You can preview 72a8997 at https://pr40427-72a8997.ngbuilds.io/.

Splaktar commented on aio/src/styles/0-base/_typography.scss:
```diff
@@ -188,8 +199,9 @@ code {
 
 .sidenav-content a {
   color: $blue;
```
The `$blue` link text in these callouts doesn't meet the contrast requirements:
![Screen Shot 2021-01-20 at 00 20 06](https://user-images.githubusercontent.com/3506071/105130626-62512e80-5ab5-11eb-9ef9-8edb2f0a857c.png)

![Screen Shot 2021-01-20 at 00 19 51](https://user-images.githubusercontent.com/3506071/105130628-62512e80-5ab5-11eb-80e9-24a2fae2765f.png)

gkalpak replied:
I couldn't find a color that has enough contrast with the blue-ish background of the `.is-helpful` callouts and at the same time is distinguishable from the regular text. (Same for the backgrounds of the other types of callouts (`.is-important`, `.is-critical`).)

Given that these colors were not changed as part of this PR, I cosider this to be outside its scope. We can fix this on a separate PR. Do you have any ideas on how we could address this?

Splaktar replied:
That sounds good, we should probably open an issue for it. If you cc me on that issue, I can try to investigate/post some ideas. I had to deal with similar things for material.angularjs.org.

gkalpak replied:
I created #40657 to track this (and cc'd you).

Splaktar commented on aio/src/styles/2-modules/_card.scss:
```diff
@@ -2,13 +2,14 @@
 
     .docs-card {
-      @include card(194px, 30%);
+      @include card(194px, 35%);
```
While you are in here, can we do something to make these cards more attractive?

Their current non-hovered state is so plain and dull. Nothing indicates that they are clickable (likely poor UX for many people):
![Screen Shot 2021-01-20 at 00 25 42](https://user-images.githubusercontent.com/3506071/105131094-426e3a80-5ab6-11eb-8a84-4d113ebe9406.png)

They look a bit more engaging with their hover styles and the styles are not so strong that all of the cards couldn't have those same styles. Can we make this coloring apply to non-hovered cards and have the hover only affect the elevation (box-shadow)?
![Screen Shot 2021-01-20 at 00 26 08](https://user-images.githubusercontent.com/3506071/105131124-4f8b2980-5ab6-11eb-8d7b-3e349acbe3a0.png)
With the colors, the cards look a lot more 'clickable'.

gkalpak replied:
Good call! Done.

Splaktar reviewed (changes requested):
I found a couple of a11y contrast issues that we need to fix.

I also have a suggestion to make the docs landing page have better UX by making the cards have a better indication that they are clickable/actionable.

Thank you so much for moving this important work forward! This is really needed to make these docs more readable (a critical feature of documentation!).

Also thank you for fixing that docs landing page card layout issue! 👏🏼

aikithoughts reviewed (approved):
Reviewed-for: global-docs-approvers

mary-poppins commented:
You can preview c7d6dbd at https://pr40427-c7d6dbd.ngbuilds.io/.

mary-poppins commented:
You can preview 0543fe7 at https://pr40427-0543fe7.ngbuilds.io/.

petebacondarwin commented on aio/src/styles/0-base/_typography.scss:
```diff
@@ -6,59 +6,70 @@ body {
+  margin-top: 4rem;
+  padding-top: 4rem;
+
+  .page-about &,
+  .page-api &,
+  .page-contribute &,
+  .page-events &,
+  .page-features &,
+  .page-home &,
+  .page-presskit &,
+  .page-resources & {
+    border-top: 0;
+    margin-top: 2rem;
+    padding-top: 0;
+  }
```
I feel a little conflicted about having these `.page-xxx` overrides dotted around in different files. I wonder if (in a future PR) we should consolidate them into a single `_page-overrides.scss` file or something?

gkalpak replied:
Yeah, that bugs me too. I have it on my list of things to try to clean up in a follow-up PR. Unfortunately, having a `_page-overrides.scss` file is not ideal either (since the overrides are disconnected from the styles they are overriding).

For now, I consider duplicating the class selectors in different places to be the lesser of two evils.

petebacondarwin reviewed (approved):
Dogged work @gkalpak

Splaktar reviewed (approved):
LGTM

sjtrimble commented:
@googlebot I consent

gkalpak committed:
fix(docs-infra): update the sidenav background to better distinguish from main content

Previously, we used a box-shadow to indicate that the sidenav is
separate from the main content.

This commits updates the sidenav styling to remove the shadow and give
it a slightly different background color instead.

Co-authored-by: Stefanie Fluin <sjtrimble@gmail.com>

gkalpak committed:
fixup! fix(docs-infra): update the sidenav background to better distinguish from main content

gkalpak committed:
fix(docs-infra): remove uppercase style from sidenav nav-items

Co-authored-by: Stefanie Fluin <sjtrimble@gmail.com>

gkalpak committed:
fix(docs-infra): fix padding in sidenav and ensure all nav-items have equal height

This commit improves the padding nav-menus and nav-items in the sidenav.
It also ensures that all nav-items have the same height, regardless of
their level and of whether they have children or not.

Co-authored-by: Stefanie Fluin <sjtrimble@gmail.com>

gkalpak committed:
fix(docs-infra): improve heading styles

This commit includes some general improvements to heading styles
(font-weight, margin/padding). It also adds a horizontal separator above
`<h2>` headings to segment important sections more easily.
(As a consequence, we can remove several `<hr />` elements right before
`<h2>` headings.)

Co-authored-by: Stefanie Fluin <sjtrimble@gmail.com>

gkalpak committed:
fixup! fix(docs-infra): improve heading styles

gkalpak committed:
fix(docs-infra): make content links more prominent on hover

Co-authored-by: Stefanie Fluin <sjtrimble@gmail.com>

gkalpak committed:
fix(docs-infra): improve typography styles (incl. bigger font size/line height)

Co-authored-by: Stefanie Fluin <sjtrimble@gmail.com>

gkalpak committed:
fixup! fix(docs-infra): improve legibility and focus on content

gkalpak committed:
fixup! fix(docs-infra): improve typography styles (incl. bigger font size/line height)

gkalpak committed:
fix(docs-infra): improve legibility and focus on content

This commit changes the layout of the main content to improve
legibility, especially on wider screens:

- Limit main content width to 50em (800px by default).
- Center main content on the page.

Co-authored-by: Stefanie Fluin <sjtrimble@gmail.com>

gkalpak committed:
refactor(docs-infra): simplify the CSS rules for styling the "collapsed/expanded" sidenav icons

This commit simplifies the styling of the "collapsed/expanded" icons in
sidenav nav-items with children by consolidating the CSS rules in one
block (instead of having duplicate blocks for different levels).

gkalpak committed:
refactor(docs-infra): clean up styles in `_card.scss`

This commit cleans up the styles in `_card.scss` by:
- Fixing incosistenct indentation/newlines.
- Removing rules for unused selectors.
- Removing unnecessary styles.

gkalpak committed:
refactor(docs-infra): remove unused styles for `.heading` CSS class

This commit removes some styles for the `.heading` CSS class which do
not have any effect (either because they are overridden in other rules
or because they do not make any difference).

NOTE:
The `.heading` class is kept in the HTML to make it easier to associate
the `.heading-children` element with its corresponding `.heading`
parent.

gkalpak committed:
fix(docs-infra): fix the styling of the cards in docs introduction page

This commit fixes the layout and appearance of the cards shown in the
docs introduction page (`/docs`) in the following ways:

- Center the cards.
- Ensure two cards are shown per line (instead of 3 one the first line
  and 1 on the second).
- Adjust their widths to ensure their content fits well in them
  (given that the cards have a fixed height).
- Use more engaging styles to better indicate that the cards are
  clickable (as discussed [here][1]).

[1]: https://github.com/angular/angular/pull/40427#discussion_r560688953

gkalpak committed:
fixup! fix(docs-infra): fix the styling of the cards in docs introduction page

gkalpak commented:
FYI: Rebased on master.

mary-poppins commented:
You can preview 41cf56b at https://pr40427-41cf56b.ngbuilds.io/.
